### PR TITLE
Ensure full page errors in new UI integration selection session are not swallowed by "Already Rendered!" error

### PR DESCRIPTION
**Pull request by SammyIsConfused**
##### Checklist

- [x] the [contributor license agreement][] is signed
- [x] commit message follows [commit guidelines][]

##### Description of change
This is a fix for two related issues. 
**Issue 1:** Errors that occur within an integration selection session are "swallowed" by another error that reads "Already Rendered!" rather than showing the correct message. This is because the IntegrationExceptionHandler doesn't rethrow like the DefaultExceptionHandler does, causing the error page to be rendered but not served up. Then when it tries to render the actual page to serve up, it triggers the "Already Rendered!" error. 

In the old UI it appears to use the Default handler anyway, and so things work properly. 

I was able to edit the IntegrationExceptionHandler so that if in the new UI the error gets properly rethrown as a RuntimeException, rather than swallowed up.

**Issue 2:** An error that was hidden by this was a known bug in the Canvas integration, in which if you rapidly click the Save button in a selection session it messes up the cookie cache for that particular call by adding in cookies repeatedly, as the cookie cache is URL specific but not request specific. This cookie cache is then used to populate a request to Canvas, but ends up being too long for Canvas's Apache to understand. Canvas' Apache spits back an error message (annoyingly as a HTML page).  

In the old UI this is impossible to do, as there is a JS error message that pops up with "This request is already being submitted", and so you can't send them fast enough to make the cookie cache get swamped.

The problem with this is, if the cookie cache gets into a malformed state like this, it will continue to occur with further saves, even after a refresh. So I've made sure to catch this error, and invalidate the cache when it occurs. This at least ensures that you can simply try again. 

**TL;DR**
- Stopped "Already Rendered!" message from getting in the way of actual error messages within Integration Selection Sessions
- Made sure that if the Canvas Integration cookie cache gets swamped it can clean itself up
- Small bit of CSS to make the error message display nicely

**Note:** I haven't fixed the cookie cache, I.E prevented it from getting swamped in the first place. All I've done is ensure that if it does get swamped, it can invalidate itself. 

**Screengrab clips**

*New UI Before fix:*

Trigger the Canvas cookie cache problem via spamming the Save button and see Already Rendered. Refresh and see that it happens from then on.
https://user-images.githubusercontent.com/24543345/107714795-fd8a9c00-6d21-11eb-9b73-410b2161f079.mp4

With SEARCH_PAGE ACL disabled, attempt to open the search page in a selection session and see Already Rendered
https://user-images.githubusercontent.com/24543345/107714891-2d39a400-6d22-11eb-9ff6-1af54a7eebd2.mp4

*New UI After fix:*

With SEARCH_PAGE ACL disabled, attempt to open the search page in a selection session and see the error that we should see.
https://user-images.githubusercontent.com/24543345/107714939-480c1880-6d22-11eb-8825-ba86169fa24d.mp4

Trigger the Canvas cookie cache problem via spamming the Save button, see the new error message. Refresh and see that the cache has been invalidated and the bug no longer happens.
https://user-images.githubusercontent.com/24543345/107715017-7984e400-6d22-11eb-8a46-ac371d41dc56.mp4

*Old UI Before and After fix:* 

Note that it is impossible to spam click the Save button in the old UI, so the canvas bug cannot occur. Appears identical before and after fix.
https://user-images.githubusercontent.com/24543345/107715240-ec8e5a80-6d22-11eb-95df-7738a22086b1.mp4

Old UI - Shows what the SEARCH PAGE ACL error message looks like. Appears identical before and after fix. 
https://user-images.githubusercontent.com/24543345/107715251-f748ef80-6d22-11eb-9efd-a0947a81a1fd.mp4

[contributor license agreement]: https://www.apereo.org/node/676
[commit guidelines]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by PenghaiZhang** [comment]
As far as I know, errors can be also swallowed when not in an integration so I hope the fix can cover that situation.

**Comment by PenghaiZhang** [comment]
Another way worth to try is to override method 'execute' in LegacyContentController.

**Comment by SammyIsConfused** [comment]
> As far as I know, errors can be also swallowed when not in an integration so I hope the fix can cover that situation.

I can't see how. The IntegrationExceptionHandler is the one that doesn't rethrow. But, if we were ever to find a situation in which this happens (I haven't found one yet and have been looking) we'd know to look at the handlers.

**Comment by edalex-ian** [comment]
Could you please add a video?

File: Source/Plugins/Core/com.equella.core/src/com/tle/web/integration/IntegrationExceptionHandler.java [hunk 1]
```diff
@@ -44,9 +46,14 @@ public boolean canHandle(SectionInfo info, Throwable ex, SectionEvent<?> event)
   @Override
   public void handle(
       Throwable exception, SectionInfo info, SectionsController controller, SectionEvent<?> event) {
+    if (RenderNewTemplate.isNewUIEnabled()) {
+      SectionUtils.throwRuntime(exception);
+    }
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/src/com/tle/web/integration/IntegrationExceptionHandler.java:51]: Interesting some of this hasn't been done in the super. Hopefully we're not repeating in too many places.

**Review by edalex-ian — COMMENTED** [review]
Looks good @SammyIsConfused , just a couple of bits of tidy up and queries.

Also will review the video once available.

File: Source/Plugins/Core/com.equella.core/src/com/tle/core/services/impl/HttpServiceImpl.java [hunk 2]
```diff
@@ -217,6 +217,14 @@ private ResponseImpl getWebContentPrivate(
     }
   }
 
+  public String toCacheKey(String url) {
+    return /* req.getSession().getId() */ "FIXME" + ':' + url;
```
[the remark this answers is not in the document]
Comment (SammyIsConfused) [reply Source/Plugins/Core/com.equella.core/src/com/tle/core/services/impl/HttpServiceImpl.java:221]: Yeah, it seems like they wanted to refer to the session ID when setting cache keys but the request object doesn't have access to that. I looked into it briefly but couldn't get it to work properly.

File: Source/Plugins/Core/com.equella.core/src/com/tle/web/integration/IntegrationExceptionHandler.java (hunk 1 above)
Comment (SammyIsConfused) [reply Source/Plugins/Core/com.equella.core/src/com/tle/web/integration/IntegrationExceptionHandler.java:51]: Yeah, theres about five or six different handlers, this one is very similar to the super. But most of the others (ajaxHandler, etc) do things wildly differently.

File: Source/Plugins/Core/com.equella.core/src/com/tle/core/services/impl/HttpServiceImpl.java [hunk 3]
```diff
@@ -217,6 +217,14 @@ private ResponseImpl getWebContentPrivate(
...
+  }
+
+  public void clearCookieCache(String url) {
```
[the remark this answers is not in the document]
Comment (SammyIsConfused) [reply Source/Plugins/Core/com.equella.core/src/com/tle/core/services/impl/HttpServiceImpl.java]: I figured out that with enough spamming you can get it to occur after this happens, unless I invalidate the whole cache (context for this cache is Canvas specific for this user, this isn't clearing caches unrelated to this session). It's harder to make it happen but still possible. So I've reworked the method to invalidate all, rather than the single entry and I can't get it to break now.

File: Source/Plugins/Core/com.equella.core/src/com/tle/web/integration/IntegrationExceptionHandler.java (hunk 1 above)
Comment (SammyIsConfused) [reply Source/Plugins/Core/com.equella.core/src/com/tle/web/integration/IntegrationExceptionHandler.java:51]: > Could you please add a video?

Videos are now up @edalex-ian :+1:

File: Source/Plugins/Core/com.equella.core/src/com/tle/core/services/impl/HttpServiceImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/src/com/tle/core/services/impl/HttpServiceImpl.java:221]: Ahhh okay, I see. Drats.

File: Source/Plugins/Core/com.equella.core/resources/com/tle/core/i18n/service/impl/i18n-resource-centre.properties [hunk 4]
```diff
@@ -401,3 +401,4 @@ institutions.error.pagenotfound = Page not found
 
 institutions.server.settings.name = Server settings
 com.tle.core.connectors.canvas.error.prelim=An error occurred in Canvas\:
+com.tle.core.connectors.canvas.error.apache=An error occurred in Canvas that was returned from Apache. Cleared cookie cache.
```
[the remark this answers is not in the document]
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/resources/com/tle/core/i18n/service/impl/i18n-resource-centre.properties:404]: Curious... In your videos this string was updated, but here in your branch it is not. Did you forget to push something?

**Review by edalex-ian — COMMENTED** [review]
Looks like it's almost there, but... did you forget to push stuff?

File: Source/Plugins/Core/com.equella.core/resources/com/tle/core/i18n/service/impl/i18n-resource-centre.properties (hunk 4 above)
[the remark this answers is not in the document]
Comment (SammyIsConfused) [reply Source/Plugins/Core/com.equella.core/resources/com/tle/core/i18n/service/impl/i18n-resource-centre.properties]: D'oh! I did, I had the phrase "git push" sat in my terminal without being run. The only thing in the one unpushed commit was the string change.

**Review by edalex-ian — APPROVED** [review]
